File: 0033-Search-in-Rotated-Sorted-Array/lc0033.cc
// C++ solution.

// Use binary search directly.
class Solution {
public:
    int search(vector<int>& nums, int target) {
        auto adjust = [&](int x) -> int {
            if (target < nums[0]) {
                return x < nums[0] ? x : numeric_limits<int>::min();
            } else {
                return x < nums[0] ? numeric_limits<int>::max() : x;
            }
        };

        int low = 0;
        int high = nums.size();
        while (low < high) {
            int mid = low + (high-low) / 2;
            int mid_val = adjust(nums[mid]);
            if (mid_val > target) {
                high = mid;
            } else if (mid_val < target) {
                low = mid + 1;
            } else {
                return mid;
            }
        }
        return -1;
    }
};

// FP style.
class Solution {
public:
    int search(vector<int>& nums, int target) {
        auto skip_left = [&](int x) -> int { return x >= nums[0] ? numeric_limits<int>::min() : x; };
        auto skip_right = [&](int x) -> int { return x < nums[0] ? numeric_limits<int>::max() : x; };
        auto adjust = [&](int x) -> int { return target < nums[0] ? skip_left(x) : skip_right(x); };
        auto comp = [&](int x, int y) -> bool { return adjust(x) < adjust(y); };
        auto it = lower_bound(nums.begin(), nums.end(), target, comp);
        return (it != nums.end() && *it == target) ? it-nums.begin() : -1;
    }
};

File: 0033-Search-in-Rotated-Sorted-Array/lc0033.rs
// Rust solution.

use std::cmp::Ordering;
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let adjust = |x: i32| -> i32 {
            match (target < nums[0]) {
                true => if x < nums[0] { x } else { std::i32::MIN },
                false => if x < nums[0] { std::i32::MAX } else { x },
            }
        };

        let (mut low, mut high) = (0_usize, nums.len());
        while low < high {
            let mid = (low + high) >> 1;
            match adjust(nums[mid]).cmp(&target) {
                Ordering::Less => { low = mid+1; },
                Ordering::Greater => { high = mid },
                Ordering::Equal => { return mid as i32; }
            }
        }
        -1
    }
}
